### PR TITLE
ceph: revert to lvm mode for host-based OSDs

### LINKED ISSUE #7940: unexpected partition on disks >= 1TB (atari partitions)
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**

```
vdb    252:16   0    3T  0 disk 
├─vdb2 252:18   0   48G  0 part 
└─vdb3 252:19   0  6.1M  0 part 
```

**Expected behavior:**

```
vdb    ceph_bluestore
```

**How to reproduce it (minimal and precise):**

1. An empty data disk >= 1TiB
2. Prepare disk, using following script https://github.com/rook/rook/blob/master/Documentation/ceph-teardown.md#zapping-devices
until now there still no child partitions in vdb, looks good.
3. Follow [Quickstart.md](https://github.com/rook/rook/blob/master/Documentation/ceph-quickstart.md)
4. Then there will be 2 OSDs on this disk, on in vdb, another in vdb2 which size is 48GiB.

### Workaround [added by @BlaineEXE]
If users wish to work around this issue, my latest suggestion is to partition any device greater than 875GB with a GPT partition table and create a single partition spanning the whole device. The GPT partition at the beginning of the disk should stop Linux from observing any "phantom" Atari partitions, and users will not have to destroy raw disk OSDs later to be re-created as LVM.

**Environment**:
* OS (e.g. from /etc/os-release): Ubuntu 20.04
* Kernel (e.g. `uname -a`):  Linux master-11 5.4.0-73-generic #82-Ubuntu SMP Wed Apr 14 17:39:42 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
* Rook version (use `rook version` inside of a Rook Pod): chart version 1.6.2
* Storage backend version (e.g. for ceph do `ceph -v`): v15.2.11
* Kubernetes version (use `kubectl version`): k3s+1.19

### PR DESCRIPTION
**Pull request by BlaineEXE**
All other attempts to fix phantom Atari partitions have had issues. We
need to patch ceph-volume to fix the issue wholistically, and while
ceph-volume does not have a fix present, we will revert to using LVM
mode which is not affected by phantom Atari partitions.

Fixes #7940

Signed-off-by: Blaine Gardner <blaine.gardner@redhat.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: pkg/daemon/ceph/osd/volume.go [hunk 1]
```diff
@@ -541,6 +541,12 @@ func (a *OsdAgent) useRawMode(context *clusterd.Context, pvcBacked bool) (bool,
 		useRawMode = false
 	}
 
+	// TODO: when we re-enable raw mode, we cannot merely remove the below lines. we must handle
+	// device-specific configs which this function doesn't take into account.
+	// see: https://github.com/rook/rook/pull/8316#issuecomment-880816972
+	logger.Debugf("won't use raw mode to avoid possibility of phantom atari partitions per Rook issue #7940")
+	useRawMode = false
+
```
Comment (BlaineEXE) [inline pkg/daemon/ceph/osd/volume.go:549]: This is the effective change for this PR.

The rest of the changes remove unnecessary/non-useful code, and there are documentation updates.

File: Documentation/ceph-common-issues.md [hunk 2]
```diff
@@ -988,44 +976,32 @@ You can see https://github.com/rook/rook/issues/7940 for more detailed informati
...
-1. First, remove the 3 OSDs associated with `/dev/sdb`, `/dev/sdb1`, and `/dev/sdb2`.
-2. Then wipe clean `/dev/sdb`.
-3. After this, allow the Rook operator to deploy a new OSD to `/dev/sdb`; this will allow Ceph to
+If you are using Rook v1.6, you must first update to v1.6.8 or higher to avoid further incidents of
+OSD corruption caused by these Atari partitions.
+
+An old workaround suggested using `deviceFilter: ^sd[a-z]+$`, but this still results in unexpected
+partitions. Rook will merely stop creating new OSDs on the partitions. It does not fix a related
+issue that `ceph-volume` that is unaware of the Atari partition problem. Users who used this
+workaround are still at risk for OSD failures in the future.
```
Comment (travisn) [inline Documentation/ceph-common-issues.md:985]: What is the risk of failure besides an extra partition showing up? The corruption will only come if the additional OSD is created on the atari partition, right?

File: Documentation/ceph-common-issues.md [hunk 3]
```diff
@@ -988,44 +976,32 @@ You can see https://github.com/rook/rook/issues/7940 for more detailed informati
...
+To resolve the issue, immediately update to v1.6.8 or higher. After the update, no corruption should
+occur on OSDs created in the future. Next, to get back to a healthy Ceph cluster state, focus on one
+corruped disk at a time and [remove all OSDs on each corrupted disk](ceph-osd-mgmt.md#remove-an-osd)
+one disk at a time.
+
+As an example, you may have `/dev/sdb` with two unexpected partitions (`/dev/sdb2` and `/dev/sdb3`)
+as well as a second corrupted disk `/dev/sde` with one unexpected partition (`/dev/sde2`).
+1. First, remove the OSDs associated with `/dev/sdb`, `/dev/sdb2`, and `/dev/sdb3`. There might be
+   only one, or up to 3 OSDs depending on how your system was affected. Again see the
+   [Rook docs](ceph-osd-mgmt.md#remove-an-osd).
```
Comment (travisn) [inline Documentation/ceph-common-issues.md:996]: nit
```suggestion
   [OSD management doc](ceph-osd-mgmt.md#remove-an-osd).
```

File: Documentation/ceph-common-issues.md [hunk 4]
```diff
@@ -988,44 +976,32 @@ You can see https://github.com/rook/rook/issues/7940 for more detailed informati
...
+2. Use `dd` to wipe the first sectors of the partitions followed by the disk itself. E.g.,
+   - `dd if=/dev/zero of=/dev/sdb2 bs=1M`
+   - `dd if=/dev/zero of=/dev/sdb3 bs=1M`
+   - `dd if=/dev/zero of=/dev/sdb bs=1M`
+3. Then wipe clean `/dev/sdb` to prepare it for a new OSD. We suggest `sgdisk --zap-all /dev/sdb`.
```
Comment (travisn) [inline Documentation/ceph-common-issues.md:1001]: Shall we link to `ceph-teardown.md#zapping-devices`?

**Review by travisn — CHANGES_REQUESTED** [review]

File: Documentation/ceph-common-issues.md (hunk 2 above)
Comment (BlaineEXE) [reply Documentation/ceph-common-issues.md:985]: No. See the explanation here: https://github.com/rook/rook/issues/7940#issuecomment-880958164

**Comment by BlaineEXE** [comment]
Closing in favor of #8319

**Pull request closed without merging**